proto/regen/ecocredit/v1alpha1/tx.proto

```diff
@@ -0,0 +1,167 @@
+    // SetPrecision allows the number of allowed decimal places to be increased, effectively making the supply more
+    // granular without actually changing any balances. It allows asset issuers to be able to issue an asset without needing
+    // to think about how many subdivisions are needed upfront. While it may not be relevant for credits which likely have
+    // a fairly stable market value, I wanted to experiment a bit and this serves as a proof of concept for a broader
+    // bank redesign where say for instance a coin like the ATOM or XRN could be issued in its own units rather than
+    // micro or nano-units. Instead an operation like SetPrecision would allow trading in micro, nano or pico in the future
+    // based on market demand. Arbitrary, unbounded precision is not desirable because this can lead to spam attacks (like
+    // sending 0.000000000000000000000000000001 coins). This is effectively fixed precision so under the hood it is still
+    // basically an integer, but the fixed precision can be increased so its more adaptable long term than just an integer.
+    rpc SetPrecision(MsgSetPrecisionRequest) returns (MsgSetPrecisionResponse);
+}
+
+// MsgCreateClassRequest is the Msg/CreateClass request type.
+message MsgCreateClassRequest {
+
+    // designer is the address of the account which designed the credit class. The designer has special permissions
+    // to change the list of issuers and perform other administrative operations.
+    string designer = 1;
+
+    // issuers are the addresses of the approved issuers.
+    repeated string issuers = 2;
+
+    // metadata is any arbitrary metadata to attached to the credit class.
+    bytes metadata = 3;
+}
+
+// MsgCreateClassResponse is the Msg/CreateClass response type.
+message MsgCreateClassResponse {
+
+    // class_id is the unique ID of the newly created credit class.
+    string class_id = 1;
+}
+
+// MsgCreateBatchRequest is the Msg/CreateBatch request type.
+message MsgCreateBatchRequest {
+
+    // issuer is the address of the batch issuer.
+    string issuer = 1;
+
+    // class_id is the unique ID of the class.
+    string class_id = 2;
+
+    // issuance are the credits issued in the batch.
+    repeated BatchIssuance issuance = 3;
+
+    // metadata is any arbitrary metadata to attached to the credit batch.
+    bytes metadata = 4;
+
+    // BatchIssuance represents the issuance of some credits in a batch to a single recipient.
+    message BatchIssuance {
+
+        // recipient is the account of the recipient.
+        string recipient = 1;
+
+        // tradable_units are the units of credits in this issuance that can be traded by this recipient.
+        // Decimal values are acceptable.
+        string tradable_units = 2;
+
+        // retired_units are the units of credits in this issuance that are effectively retired by the issuer on receipt.
+        // Decimal values are acceptable.
+        string retired_units = 3;
```

Comment:
Why issuing _retired_ credits? We can keep it for flexibility, but I'm curious if this makes sense.

Reply:
Also, shouldn't we use `oneof` for tradeable and retired units / credits?

Reply:
It's a fundamental part of the use case. It's not mutually exclusive either. You could issue both. In some cases it is desirable or even required to issue credits as retired (i.e. non-tradable) on creation